File: compat/suites/python-sdk/lib/scenario/loader.py
"""
lib/scenario/loader.py — reading the scenario IR files the registry names.

A generated registry group carries a ``scenario`` field: a repository-relative
path such as ``compat/model/scenarios/sqs.json``. It is resolved from the same
base the loader resolves ``registry.json`` from — never from the current
working directory, because the suite is run from the repo root by
``cmd/compat``, from ``compat/suites/python-sdk`` by hand, and from ``/`` in a
container.

Files are read lazily and once: the pilot is two files, but a service is only
opened when a group of that service is actually built.
"""

from __future__ import annotations

import json
import os
import sys
import threading
from dataclasses import dataclass
from typing import Any, Optional

from ..registry import REPO_ROOT


@dataclass(frozen=True)
class ScenarioGroup:
    """One group of one scenario file, with the client spec it is executed
    against and the file it came from (for failure messages)."""

    file: str  # repository-relative, exactly as the registry spells it
    service: str
    client: dict
    name: str
    kind: str
    setup: list[dict]
    tests: dict[str, dict]
    teardown: list[dict]


class ScenarioLibrary:
    """Lazily loads scenario files and indexes their groups by name.

    Thread-safe: groups are resolved while the registry is being built (one
    thread), but the cache is shared with nothing that would notice a race
    either way, and a lock costs nothing here."""

    def __init__(self, base_dir: str = REPO_ROOT) -> None:
        self._base_dir = base_dir
        self._lock = threading.Lock()
        self._files: dict[str, Optional[dict[str, ScenarioGroup]]] = {}

    def group(self, scenario_file: Optional[str], group_name: str) -> Optional[ScenarioGroup]:
        """The named group of that scenario file, or None when there is no
        such file or no such group — "not mine", which is what the backend
        hook returns for anything it cannot resolve."""
        if not scenario_file:
            return None
        with self._lock:
            groups = self._load(scenario_file)
        if groups is None:
            return None
        return groups.get(group_name)

    def _load(self, scenario_file: str) -> Optional[dict[str, ScenarioGroup]]:
        if scenario_file in self._files:
            return self._files[scenario_file]

        path = os.path.join(self._base_dir, *scenario_file.split("/"))
        groups: Optional[dict[str, ScenarioGroup]]
        try:
            with open(path, encoding="utf-8") as f:
                data = json.load(f)
            groups = _index(scenario_file, data)
        except (OSError, ValueError, KeyError, TypeError) as exc:
            # Not fatal, and not silent. The registry and the scenario files
            # are generated together, so this cannot happen in a healthy tree;
            # when it does, returning None lets the loader's own sentinel
            # report a loud failure for every test of the group rather than
            # aborting the whole suite over one unreadable file.
            sys.stderr.write(
                f"[compat:python-sdk] cannot read scenario file {path}: {exc}\n"
            )
            groups = None

        self._files[scenario_file] = groups
        return groups


def _index(scenario_file: str, data: Any) -> dict[str, ScenarioGroup]:
    version = data["version"]
    if version != 1:
        raise ValueError(f"unsupported scenario version {version!r} (want 1)")
    service = data["service"]
    client = data["client"]
    indexed: dict[str, ScenarioGroup] = {}
    for group in data["groups"]:
        indexed[group["name"]] = ScenarioGroup(
            file=scenario_file,
            service=service,
            client=client,
            name=group["name"],
            kind=group["kind"],
            setup=list(group.get("setup") or []),
            tests={t["name"]: t for t in group["tests"]},
            teardown=list(group.get("teardown") or []),
        )
    return indexed

File: compat/suites/cli/internal/scenario/loader.go
// Package scenario executes the generated compat scenario IR — the files under
// compat/model/scenarios/, described normatively by compat/model/README.md — as
// cli-suite tests.
//
// It is the cli half of G2 (docs/plans/compat-coverage-modelgen.md § 3.2). A
// generated registry group carries a "scenario" path instead of a Go
// implementation; internal/registry consults Backend.Resolve for any test with
// no static impl, and Backend.Setup/Backend.Teardown supply the group hooks the
// same file declares. Every call is made the way a hand-written group makes one
// — `aws <command> <kebab-op> --cli-input-json '<json>'` through
// internal/awscli — so the interpreter exercises the CLI's own request shaping
// with no Overcast-specific code path.
package scenario

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
)

// scenarioVersion is the only scenario file version this interpreter accepts.
// A bump has to be a deliberate multi-file change, exactly as it is for
// registry.generated.json in internal/registry.
const scenarioVersion = 1

// File is one scenario file: compat/model/scenarios/<service>.json.
type File struct {
	Version int     `json:"version"`
	Service string  `json:"service"`
	Client  Client  `json:"client"`
	Groups  []Group `json:"groups"`

	// Path is the repository-relative path the registry named this file by,
	// with '/' separators. It is failure-message field 6, so it is kept
	// exactly as written rather than reconstructed from the service name.
	Path string `json:"-"`
}

// Client is what an interpreter needs to reach the service without a naming
// table of its own (compat/model/README.md § Naming). The cli backend uses
// EndpointPrefix as the `aws` command name; the remaining fields are carried so
// the type matches the IR and a later protocol-sensitive check has them.
type Client struct {
	SDKID          string `json:"sdkId"`
	EndpointPrefix string `json:"endpointPrefix"`
	SigningName    string `json:"signingName"`
	Protocol       string `json:"protocol"`
	APIVersion     string `json:"apiVersion"`
	TargetPrefix   string `json:"targetPrefix"`
}

// Group is one registry group: its setup calls, its tests and its teardown.
type Group struct {
	Name     string `json:"name"`
	Kind     string `json:"kind"`
	Setup    []Call `json:"setup"`
	Tests    []Test `json:"tests"`
	Teardown []Call `json:"teardown"`
}

// Call is one service call: an operation, its input members as value
// expressions, and the context paths its response fills in.
type Call struct {
	Op     string            `json:"op"`
	Params map[string]any    `json:"params"`
	Export map[string]string `json:"export"`
}

// Test is one registry test: a primary call and the clauses that verify it.
type Test struct {
	Name    string      `json:"name"`
	Op      string      `json:"op"`
	Call    Call        `json:"call"`
	Assert  []Assertion `json:"assert"`
	Depends []string    `json:"depends"`
}

// Assertion is one clause of the IR's closed assertion set. Kind selects which
// of the remaining fields are present; the schema
// (compat/model/scenario.schema.json) is what guarantees the combination.
type Assertion struct {
	Kind        string           `json:"kind"`
	Checks      map[string]Check `json:"checks"`
	Call        *Call            `json:"call"`
	ItemsPath   string           `json:"itemsPath"`
	Where       map[string]any   `json:"where"`
	Error       *ErrorClause     `json:"error"`
	MaxAttempts int              `json:"maxAttempts"`
	DelayMs     int              `json:"delayMs"`
	Assert      *Assertion       `json:"assert"`
}

// ErrorClause carries the modeled shape name and the wire code. SDKs disagree
// about which of the two they surface, so a match on either is accepted — see
// matchesError.
type ErrorClause struct {
	Shape string `json:"shape"`
	Code  string `json:"code"`
}

// CheckKind names the one check a Checks entry carries.
type CheckKind string

// The closed set of checks (compat/model/README.md § Assertions).
const (
	CheckNonEmpty CheckKind = "nonEmpty"
	CheckIsList   CheckKind = "isList"
	CheckEquals   CheckKind = "equals"
	CheckMatches  CheckKind = "matches"
	CheckMissing  CheckKind = "missing"
)

// Check is exactly one check against one response path. Value carries the
// `equals` value expression or the `matches` pattern; the three boolean checks
// leave it nil.
//
// It is decoded by hand rather than as a struct of optional fields because
// `{"equals": null}` is a legal check — the IR's Value admits null — and a
// struct cannot tell that from an absent key.
type Check struct {
	Kind  CheckKind
	Value any
}

// UnmarshalJSON decodes the single-entry object the schema requires.
func (c *Check) UnmarshalJSON(b []byte) error {
	var raw map[string]any
	if err := json.Unmarshal(b, &raw); err != nil {
		return err
	}
	if len(raw) != 1 {
		return fmt.Errorf("check must carry exactly one of nonEmpty/isList/equals/matches/missing, got %d entries", len(raw))
	}
	for k, v := range raw {
		switch CheckKind(k) {
		case CheckNonEmpty, CheckIsList, CheckMissing:
			c.Kind, c.Value = CheckKind(k), nil
		case CheckEquals, CheckMatches:
			c.Kind, c.Value = CheckKind(k), v
		default:
			return fmt.Errorf("unknown check %q", k)
		}
	}
	return nil
}

// group returns the named group.
func (f *File) group(name string) (*Group, bool) {
	for i := range f.Groups {
		if f.Groups[i].Name == name {
			return &f.Groups[i], true
		}
	}
	return nil, false
}

// test returns the named test.
func (g *Group) test(name string) (*Test, bool) {
	for i := range g.Tests {
		if g.Tests[i].Name == name {
			return &g.Tests[i], true
		}
	}
	return nil, false
}

// Backend loads scenario files and turns their groups into cli-suite test,
// setup and teardown functions. One Backend serves the whole run; it is safe
// for concurrent use, which it has to be because the harness runs groups in
// parallel.
//
// It holds no per-run state of its own: the context bag a group's exports write
// into lives on the harness TestContext, which is created fresh for every group
// run, so re-running a group in interactive mode starts from an empty context
// exactly as the first run did.
type Backend struct {
	// root is the repository root; a group's Scenario path is relative to it.
	root string
	// run is the process runner. Production is awscliRunner; the tests
	// substitute an in-memory fake.
	run runner

	mu    sync.Mutex
	files map[string]*File
	errs  map[string]error
}

// New returns a Backend that resolves scenario paths against root and calls the
// real `aws` binary through internal/awscli.
func New(root string) *Backend {
	return &Backend{root: root, run: awscliRunner{}, files: map[string]*File{}, errs: map[string]error{}}
}

// load reads and caches one scenario file, keyed by the repository-relative
// path the registry named. Both the file and a load failure are cached: a
// broken file must produce the same loud failure for every test in the group
// rather than a different one per read.
func (b *Backend) load(rel string) (*File, error) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if f, ok := b.files[rel]; ok {
		return f, nil
	}
	if err, ok := b.errs[rel]; ok {
		return nil, err
	}
	f, err := readFile(filepath.Join(b.root, filepath.FromSlash(rel)), rel)
	if err != nil {
		b.errs[rel] = err
		return nil, err
	}
	b.files[rel] = f
	return f, nil
}

// readFile parses one scenario file. rel is recorded on the result so failure
// messages name the file the registry pointed at, not the absolute path this
// process happened to open.
func readFile(abs, rel string) (*File, error) {
	raw, err := os.ReadFile(abs) //nolint:gosec // the path comes from the registry, not from user input
	if err != nil {
		return nil, fmt.Errorf("scenario: read %s: %w", rel, err)
	}
	var f File
	if err := json.Unmarshal(raw, &f); err != nil {
		return nil, fmt.Errorf("scenario: parse %s: %w", rel, err)
	}
	if f.Version != scenarioVersion {
		return nil, fmt.Errorf("scenario: %s: version %d, want %d", rel, f.Version, scenarioVersion)
	}
	if f.Client.EndpointPrefix == "" {
		return nil, fmt.Errorf("scenario: %s: client.endpointPrefix is empty, so no aws command name can be derived", rel)
	}
	f.Path = rel
	return &f, nil
}
